### ISSUE TITLE
[BUG]: ```__bool__()``` called on ```SIMD[DType.uint64,1]()```

### ISSUE THREAD

**Issue by rd4com** [modularml/mojo#1623]
### Bug description

Hello, there might be a small bug with the ```__bool__()``` method of ```SIMD``` non bools DTypes, 
bool seem to behave like expected, here is some code just in case it is a bug.

it might be related to: [Link for a message on the chat](https://discord.com/channels/1087530497313357884/1151418092052815884/1196095184274993223)

Numbers are all fine, it might be ```__bool__()``` called on ```SIMD[DType.uint64,1]()```

> ```__bool__``` True if all the elements in the SIMD vector are non-zero and False otherwise.
> [Documentation: ```SIMD.__bool__()```](https://docs.modular.com/mojo/stdlib/builtin/simd.html#bool__) 

```python
def main():

    if SIMD[DType.int64,1](0):
        print("OK0") #print
    if int(SIMD[DType.int64,1](0)):
        print("OK1") #Good, not print

    if SIMD[DType.int64,1](0).__bool__():
        print("OK3") #print
    if int(SIMD[DType.int64,1](0)).__bool__():
        print("OK4") #Good, not print


    # Numbers fine:   
    if (SIMD[DType.int64,1](0)&SIMD[DType.int64,1](1))>0:
        print("(a&b)>0") #✅ correct, not print
    if SIMD[DType.int64,1](0)|SIMD[DType.int64,1](1)>0:
        print("(a|b)>0") #✅ correct, print

    if (SIMD[DType.int64,1](0)&SIMD[DType.int64,1](1)):
        print("a&b") #✅print it, ❌should not
    if SIMD[DType.int64,1](0)|SIMD[DType.int64,1](1):
        print("a|b") #❌ not print, should
    

    print(Int64(0).__bool__())       #❌True
    if 1:        print("1")          #✅ print 1
    if Int64(1): print("Int64(1)")   #❌ not print

    print("1.",SIMD[DType.int64,1](99).__bool__()) #False
    print("2.",SIMD[DType.int64,1](1).__bool__()) #False
    print("3.",SIMD[DType.int64,1](0).__bool__()) #True
    print("4.",SIMD[DType.int64,1](-1).__bool__())#False
    print("5.",SIMD[DType.uint8,2](1,2).__bool__())#False
    print("6.",SIMD[DType.uint8,2](1,2).cast[DType.bool]().__bool__()) #true✅ 
    print("7.",SIMD[DType.uint8,4](15,8,4,2).__bool__()) #false
    print("8.",SIMD[DType.uint8,4](15,8,4,2).cast[DType.bool]().reduce_and()) #✅true
    print("9.",SIMD[DType.uint8,4](15,8,4,2).cast[DType.bool]().reduce_or()) #✅true
    
    if SIMD[DType.int64,1](1):
        print("print")  #❌not print
    
    if (SIMD[DType.int64,1](2)&SIMD[DType.int64,1](2)):
        print("ok")     #❌not print
    print(SIMD[DType.int64,1](2)&SIMD[DType.int64,1](2))
                        #✅2


    if (SIMD[DType.int64,1](2)&SIMD[DType.int64,1](2))>0:
        print("(a&b)>0")    #✅print
    if ((SIMD[DType.int64,1](2)&SIMD[DType.int64,1](2))):
        print("(a&b)")      #❌not print, __bool__ could be the problem
    if ((SIMD[DType.int64,1](0)&SIMD[DType.int64,1](0))):
        print("(a0&b0)")    #✅print,❌ should not, __bool__ could be the problem

    print(SIMD[DType.int64,1].__bool__(SIMD[DType.int64,1](-1))) #False
    print(SIMD[DType.int64,1].__bool__(SIMD[DType.int64,1](0))) #True
    print(SIMD[DType.int64,1].__bool__(SIMD[DType.int64,1](1))) #False

    if SIMD[DType.int64,1](2).__bool__() and SIMD[DType.int64,1](2).__bool__():
        print("__bool__() & __bool__()")  #❌

    if SIMD[DType.int64,1](0) and SIMD[DType.int64,1](0):
        print("0 and 0")  #❌

    if 15&(1<<3): print("ok1")        #✅ print
    if Int64(15)&(1<<3): print("ok2") #❌ not print
    if (Int64(15)&(1<<3))>0: print("ok3") #✅ print

    for n in range(10):
        var x = SIMD[DType.uint64,1](n)
        var res = 0
        var res2 = 0
        for i in range(64):
            var tmp = SIMD[DType.uint64,1](1)<<i
            if x&tmp: res+=1
            if (x&tmp)>0: res2+=1
        print(res,res2)
        
        #64 0
        #63 1
        #63 1
        #62 2
        #63 1
        #62 2
        #62 2
        #61 3
        #63 1
        #62 2
```

### Steps to reproduce

Nothing special, it is in the description

### System information

```shell
- Ubuntu 23.10
- mojo 0.6.1 (876ded2e)
- modular 0.2.2 (95d42445)
```

**Comment by soraros** [comment]
Put it more simply, `__bool__` seems to implement `not bool(...)` on `[float|(u)int][16|32|46]` and god knows what on `(u)int8`. The following program prints `0`.

```mojo
fn main():
  for i in range(-2**15, 2**15-1):
    if Int16(i).__bool__():
      print(i)
```

**Comment by JoeLoser** [comment]
@ConnorGray or @jackos - either of you interested in looking at this bug?

**Comment by soraros** [comment]
Seems to be fixed in 24.1.

Actually, I think, this is a duplicate of #1442.

**Comment by rd4com** [comment]
Yes, good job! :+1: 
It is all fine according to a re-run of the code of bug description with mojo nightly ```2024.4.818 (34399306)``` !

@soraros, this issue does seem to be a duplicate of #1442;

Sorry for the duplicate issue, i'll close it as resolved :partying_face: !
